internal/initializer/tls.go

```diff
@@ -0,0 +1,332 @@
+	errDecodeKey               = "cannot decode key"
+	errDecodeCert              = "cannot decode cert"
+	errFmtGetTLSSecret         = "cannot get TLS secret: %s"
+	errFmtCannotCreateOrUpdate = "cannot create or update secret: %s"
+)
+
+const (
+	// RootCACertSecretName is the name of the secret that will store CA certificates and rest of the
+	// certificates created per entities will be signed by this CA
+	RootCACertSecretName = "root-ca-certs"
+)
+
+// TLSCertificateGenerator is an initializer step that will find the given secret
+// and fill its tls.crt, tls.key and ca.crt fields to be used for External Secret
+// Store plugins
+type TLSCertificateGenerator struct {
+	namespace           string
+	caSecretName        string
+	tlsServerSecretName string
+	tlsClientSecretName string
+	subject             string
+	owner               []metav1.OwnerReference
+	certificate         CertificateGenerator
+	log                 logging.Logger
+}
+
+// TLSCertificateGeneratorOption is used to configure TLSCertificateGenerator behavior.
+type TLSCertificateGeneratorOption func(*TLSCertificateGenerator)
+
+// TLSCertificateGeneratorWithLogger returns an TLSCertificateGeneratorOption that configures logger
+func TLSCertificateGeneratorWithLogger(log logging.Logger) TLSCertificateGeneratorOption {
+	return func(g *TLSCertificateGenerator) {
+		g.log = log
+	}
+}
+
+// NewTLSCertificateGenerator returns a new TLSCertificateGenerator.
+func NewTLSCertificateGenerator(ns, caSecret, tlsServerSecret, tlsClientSecret, subject string, owner []metav1.OwnerReference, opts ...TLSCertificateGeneratorOption) *TLSCertificateGenerator {
+	e := &TLSCertificateGenerator{
+		namespace:           ns,
+		caSecretName:        caSecret,
+		tlsServerSecretName: tlsServerSecret,
+		tlsClientSecretName: tlsClientSecret,
+		subject:             subject,
+		owner:               owner,
+		certificate:         NewCertGenerator(),
+		log:                 logging.NewNopLogger(),
+	}
+
+	for _, f := range opts {
+		f(e)
+	}
+	return e
+}
+
+func (e *TLSCertificateGenerator) loadOrGenerateCA(ctx context.Context, kube client.Client, nn types.NamespacedName) (*CertificateSigner, error) {
+	caSecret := &corev1.Secret{}
+
+	err := kube.Get(ctx, nn, caSecret)
+	if resource.IgnoreNotFound(err) != nil {
+		return nil, errors.Wrapf(err, errFmtGetTLSSecret, nn.Name)
+	}
+
+	if err == nil {
+		kd := caSecret.Data[SecretKeyTLSKey]
+		cd := caSecret.Data[SecretKeyTLSCert]
+		if len(kd) != 0 && len(cd) != 0 {
+			e.log.Info("TLS CA secret is complete.")
+			return parseCertificateSigner(kd, cd)
+		}
+	}
+	e.log.Info("TLS CA secret is empty or not complete, generating a new CA...")
+
+	a := &x509.Certificate{
+		SerialNumber:          big.NewInt(2022),
+		Subject:               pkixName,
+		Issuer:                pkixName,
+		DNSNames:              []string{"crossplane-root-ca"},
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().AddDate(10, 0, 0),
```

Comment:
This means our certificates will expire in 10 years, right?

Out of scope for this PR, but have we given any thought to using shorter lived certificates?

Reply:
Agree with @negz. We should probably also take into consideration implementing certificate rotation in the same PR.

Reply:
> This means our certificates will expire in 10 years, right?

Yes and with the current implementation, certificates can easily be rotated in case of leakage or expiration by deleting their content and restarting Crossplane pod (if the certificate is created by helm). I plan to document it once this PR and following PRs are merged.

I prefer having an automated rotation in place before creating shorter lived certificates.